CenterForOpenScience/osf.io
# [Migration Required] Feature/Registered Meta Migration [ENG-842]

⭐️ Majority of this PR is a test to keep behavior from regressing.  Deeply nested registration test metadata is pretty lengthy, which comprises the bulk of the lines of this PR.

## Purpose 

Now that registration schemas have been flattened into schema blocks, the responses to these schema questions need to be likewise flattened.

## Changes

- Adds `registration_responses` fields and `registration_responses_migrated` (temporary) fields to `DraftRegistration` and `Registration` models
- Adds a migration to pull the relevant information from `DraftRegistration.registration_metadata` and `Registration.registered_meta` into `DraftRegistration.registration_responses` and `Registration.registration_responses` in local development - hoisting all nested answers to the top level. Given a `registration_response_key`, we want to be able to do `registration_responses[registration_response_key]` to get at the answer.
- On staging/prod, a management command will need to be run to incrementally populate these columns. 
- Adds `registration_responses` to the DraftRegistration and Registration serializers 
   - On the DraftRegistration serializer,  if you write to `registration_responses`, it is expanded and written to `registration_metadata`.  The reverse is also true, if you write to `registration_metadata`, it is flattened and written to `registration_responses`. 
- `Node.register_node` updated so you pass in the `draft_registration` as one of the parameters instead of the `registration_metadata`
- Registration supplemental questions pertaining to contributors are overwritten as part of the registration process with bibliographic contributors on the node.


## Example - open-ended registration
```json
"registered_meta": {
    "summary": {
           "comments": [],
           "value": "hello",
           "extra": []
      }
  },
 "registration_responses": {
    "summary": "hello"
}
```

files are stored like this:
```python
"recommended-hypothesis.question4a": [
                    {
                        "file_name": "Alphabet.txt",
                        "file_id": "5d6d25024d476c088fb8e03b",
                        "sha256": "1ac3f7ddc8792ea3e5a0db7sdfsf955d8aaabff6b394446aed708"
                    },
                    {
                        "file_name": "Screen Shot 2019-08-30 at 9.04.01 AM.png",
                        "file_id": "5d6d25014d476c088fb8e038",
                         "sha256": "sdfsdfsdfasdfasd3e5a0db7sdfsf955d8aaabff6b394446aed708"
                    }
                ]
```

## QA Notes
⭐️ Has both a migration and a management command. New fields are added to the db with the migration, but they are populated with a management command. Get devops to run once your prep is complete.
- Migration (management command prep) - make several draft registrations and several registrations with different schemas (veer, OSF preregistration, osf standard, etc.).  The management command is going to go through all of the `registered_metadata` on DraftRegistrations and all of the `registered_meta` on Registrations and populated `registration_responses` from them.   `registration_responses` fields are going to be much flatter.   Post-migration, visit the DraftRegistrations and the Registrations through the API and verify that the `registration_responses` field is populated and all of the information has been mapped to a flatter, top-level, correctly.

For example:
registration_metadata: 
```
{
    "summary": {
           "comments": [],
           "value": "hello",
           "extra": []
      }
  }
```
Turns into `registration_responses`:
```
 "registration_responses": {
    "summary": "hello"
}
```
### QA Post migration
- Create a registration through UI normally.  Verify that the API has `registration_responses` populated for that registration, which is a flattened version of the `registration_metadata`. 
- Start a draft registration through the UI normally. Verify that the API has `registration_responses` populated for the draft. 
- Create a draft registration in the API, posting to `registration_responses`.  Verify that `registration_metadata` is created for you.
- Verify that you cannot POST to both `registration_responses` and `registration_metadata` simultaneously in the API.
- If you view a registration in the API with a view-only-link, authors questions on OSF Preregistration should be anonymized in both the `registered_meta`, and the `registered_responses`
- Verify that contributor questions are overwritten when the draft is actually registered (question 2 on OSF Preregistration, for example). Through the API, update this question to be the name of the random person.  When you actually register the draft, that question will be overwritten with the names of the bibliographic contributors.
- Verify that if you create a draft registration API only (writing to `registration_responses`), and then register the draft, that only the `file_id`, `file_name`, and correct `sha256` for that file are needed for file questions.   
for example, this is one of the file questions - if these pieces are there in the draft registration, the URL for the file during registration should resolve to the file on the registration.
```
"registration_responses": {
    "q7.uploader": [
			        {
			            "file_name": "groups_react.html",
			            "file_id": "5d9260bcab42a00684747d3d",
			            "sha256": "1ac3f7ddc8792ea3e5a0db708efa1d2cf05cc3f955d8aaabff6b394446adsds"
			        }
			    ]
}
```

## Documentation



## Side Effects



## Ticket

https://openscience.atlassian.net/browse/ENG-842
